### DISCUSSION TITLE
Node.js updates from docker mirror not grouped with other node updates

### DISCUSSION THREAD

**Discussion by fabb** [renovatebot/renovate#16646] [Request Help] [answered]
This is a continuation from https://github.com/renovatebot/renovate/issues/16634.

I just use `config:base` in a repo where I use `Dockerfile`/`.nvmrc`/`gitlab-ci.yml`. In some of my Dockerfiles, a private docker mirror is used. I would expect all Node.js minor package updates to be grouped, but the ones using the docker mirror are not.

This is the issue I am seeing:

![Screenshot 2022-07-18 at 12 46 45](https://user-images.githubusercontent.com/153960/179677663-78bb56c4-bc47-48d2-abf4-86a81b72a2ce.png)

![Screenshot 2022-07-18 at 12 46 56](https://user-images.githubusercontent.com/153960/179677683-75f70644-3cf6-489c-9b6c-c88d07c51e3c.png)

So somehow renovate _does_ see the container image updates in my local docker mirror correctly, but does not group them with the other node updates. 

The second screenshot shows the packages that are grouped by the `nodeJs` rule included by `config:base`:
```
  nodeJs: {
    description:
      "Group anything that looks like Node.js together so that it's updated together.",
    packageRules: [
      {
        matchDatasources: ['docker'],
        matchPackageNames: ['node'],
        matchPackagePatterns: ['/node$'],
        excludePackageNames: ['calico/node', 'kindest/node'],
        commitMessageTopic: 'Node.js',
      },
    ],
  },
```
https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/group.ts#L34-L46

But why does it not cover the containers referenced via docker mirror shown in the first screenshot? Shouldn't `matchPackagePatterns: ['/node$']` correctly match that?

**Comment by rarkins** [comment]
What are the two branch names?

**Reply by fabb** [reply]
* First image: `renovate/hub.myprivatemirror.com-8448-node-16.x` (8448 is the port)
* Second image: `renovate/node-16.x`

**Comment by rarkins** [comment]
You could try writing a package rule to set groupName=node for all the packages you want to match (including your custom registry one)

**Reply by fabb** [reply]
I will try that, thanks. But do you have an idea why the built-in `nodeJs` rule does not already match the custom registry node package? I‘m trying to understand the config and I don‘t see how `matchPackagePatterns: ['/node$']` does not match.

**Reply by PhilipAbed** [reply]
why the slash at start?

**Reply by fabb** [reply]
I guess to match mirrors. This is form the renovate `config:base`, see initial post.

**Reply by rarkins** [reply]
The rule changes the commit message but not branch?

**Reply by fabb** [reply]
Not sure what you mean?

**Reply by PhilipAbed** [reply]
he means that the Rule is catching both cases, and is adding a commit message of `Update NodeJs`, 
but the other Message is changing to `Update NodeJs to v16.16.0`
but what about the branch, Every pull request is connected to a branch, 
is it also using a different branch or the same branch for 2 pull requests?

**Reply by PhilipAbed** [reply]
if both the `updateType` are `Minor` then it could be the hardcoded section which is causing this

https://github.com/renovatebot/renovate/blob/a6b747444a4ac4df9e6841aa5b21dbcb5b816b6c/lib/modules/manager/npm/extract/index.ts#L512-L516

that means that dependency names that are `node` will be alone, and others which have anything else with `XYZ/node` will be in a different pr

**Reply by fabb** [reply]
Ah good catch. I thought the rule was only catching the second PR, and the first PR was just created by the default renovate behavior.

So is it a bug? Can we reopen and rephrase the ticket https://github.com/renovatebot/renovate/issues/16634?

**Reply by rarkins** [reply]
Grouping depends on branch names and not commit messages. The rule modified commit message and not group name

**Comment by PhilipAbed** [comment]
@fabb what are the branchnames?

**Reply by fabb** [reply]
See this message: https://github.com/renovatebot/renovate/discussions/16646#discussioncomment-3179066

**Reply by PhilipAbed** [reply]
i tried to take your repository from the closed issue to reproduce, 
i see it has 
![image](https://user-images.githubusercontent.com/42116482/179988105-353e4888-a29c-467b-88ab-d3936dd6a009.png)
2 different Branches from what you showed me, 
both were called `Node-{major}.X`.

i added `allowedVersions: '< 17.0.0'` to the rules because your reproduction was trying to get npm 18+ instead of npm 16...
but im unable to reproduce 

im getting this: 
```
DEBUG: Found no results from datasource that look like a version (repository=StinkyLord/renovate-node-grouping, dependency=ghcr.io/docker-mirror/node)
       "result": {"registryUrl": "https://ghcr.io", "releases": []}

```

**Reply by PhilipAbed** [reply]
since the name is `node-18.x` i suppose it's gonna get grouped up if it was named `node-16.x`
my hunch  is that the `renovate/` prefix is causing the different branch name

**Reply by fabb** [reply]
I could not reproduce the issue with the private docker mirror in a github repo, so please take the example project with a grain of salt. Do you have a docker mirror that i could use for the example project?

**Comment by PhilipAbed** [comment]
@fabb reproduction repository 
https://github.com/StinkyLord/renovate-node-grouping

please confirm

**Reply by fabb** [reply]
Yes, exactly!

Good idea to reproduce it with `cimg/node`, didn't know that.

**Comment by PhilipAbed** [comment, marked as answer]
i dont think the rule is grouping them by any means.
it just says im adding A commit note,
and if you look at the names they both have `Node.js` on them

in order to group them u need to add: `"groupName": "nodeJs"`
take a look
https://github.com/StinkyLord/renovate-node-grouping/blob/main/renovate.json5

and after running i got:
https://github.com/StinkyLord/renovate-node-grouping/pull/8

`node` and `cimg/node` are 2 different dependencies
as long as they have different branch name they won't be grouped until you specify it in your package rules.
you can simply copy the same rule from config:base, and add to it.. it will override ( last matching package rule wins)

**Reply by fabb** [reply]
Ok that‘s great! So the rule itself didn‘t group, but renovate grouped automatically because the dependency was called the same for multiple changes?

What‘s this for in the example?
`"allowedVersions": "<16.17.0"`

**Reply by PhilipAbed** [reply]
for reproduction purposes only, if i dont do this, i get an update for a major version. this keeps me in minor to reproduce, read documentation

**Reply by PhilipAbed** [reply]
if that's the answers your discussion please mark it as Answered.

**Reply by fabb** [reply]
Thanks!
